addons/cert.py

```diff
@@ -234,6 +245,101 @@ def msc30(data):
         if simpleMatch(token, "rand ( )") and isStandardFunction(token):
             reportError(token, 'style', 'Do not use the rand() function for generating pseudorandom numbers', 'MSC30-c')
 
+# STR03-C
+# Do not inadvertently truncate a string
+def str03(data):
+    for token in data.tokenlist:
+        if token.str=='strncpy':
+            nextToken = token.astParent
+            while(nextToken):
+                if nextToken.str=='sizeof':
+                    reportError(nextToken, 'style', 'Do not inadvertently truncate a string', 'STR03-C')
+                nextToken = nextToken.next
+
+# STR05-C
+# Use pointers to const when referring to string literals
+def str05(data):
+    for token in data.tokenlist:
+        if token.isString:
+            parent = token.astParent
+            if parent is None:
+                continue
+            parentOp1 = parent.astOperand1
+            if parent.isAssignmentOp and not parentOp1.valueType is None:
+                if (parentOp1.valueType.type =='char' or parentOp1.valueType.type =='wchar_t') and parentOp1.valueType.pointer and not parentOp1.valueType.constness:
+                    reportError(parentOp1, 'style', 'Use pointers to const when referring to string literals', 'STR05-C')
+                
+
+# STR07-C
+# Use the bounds-checking interfaces for string manipulation
+def str07(data):
+    for token in data.tokenlist:
+        if token.str=='strcpy' or token.str=='strcat' or token.str=='fputs':
+            parent = token.astParent
+            if parent is None:
+                continue
+            if not parent.astParent :
+                reportError(token, 'style', 'Use the bounds-checking interfaces for string manipulation', 'STR07-C')
+
+# STR11-C
+# Do not specify the bound of a character array initialized with a string literal
+def str11(data):
+    for token in data.tokenlist:
+        if token.isString:
+            strlen = token.strlen
+            parent = token.astParent
+            if parent is None:
+                continue
+            parentOp1 = parent.astOperand1
+            if parent.isAssignmentOp and not parentOp1.valueType is None:
```

Comment:
you should probably check that there is an initialization. check that parentOp1 is a `[` and that its op1 is a name.

as far as I see you will warn here about:

    const char *x[10]; x[3] = "abc";

Reply:
ok. I updated the tests and the checks based on your findings. The mentioned false positives should be resolved now.